### PR TITLE
[WIP] Reference mode store

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: src/runtime/storageNG/reference-mode-store.ts [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+  /*
+   * Registered callbacks to Storage Proxies
+   */
+  private callbacks: Map<number, ProxyCallback<Container>> = new Map();
+  private nextCallbackID = 1;
+
+  /*
+   * A randomly generated key that is used for synthesized entity CRDT modifications.
+   *
+   * When entity updates are recieved by instances of ReferenceModeStore, they're non-CRDT blobs of data.
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Nit: received :stuck_out_tongue:

File: src/runtime/storageNG/reference-mode-store.ts [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+
+  /*
+   * The versions dictionary tracks the maximum write version for each entity ID, to ensure synthesized
+   * updates can be correctly applied downstream.
+   */
+  private versions: Dictionary<Dictionary<number>> = {};
+
+  /*
+   * A queue of incoming updates from the backing store, container store, and connected proxies.
+   * These are dealt wit atomically, to avoid transient states where an operation has only been partially
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: wit -> with

File: src/runtime/storageNG/reference-mode-store.ts [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+   * processed (e.g. backing written but container update not written).
+   */
+  private receiveQueue: EnqueuedMessage<Container, CRDTEntityTypeRecord<S, C>, ReferenceContainer>[] = [];
+
+  /*
+   * A queue of send Runnables. Some of these may be blocked on entities becoming available in the
+  * backing store.
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: \* alignment is off

File: src/runtime/storageNG/reference-mode-store.ts [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+      if (startProcessing) {
+        void this.processQueue();
+      }
+    });
+  }
+
+  private async processQueue() {
+    while (this.receiveQueue.length > 0) {
+      const nextMessage = this.receiveQueue[0];
+      console.log('pQ processing', nextMessage);
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Do you still want this?

File: src/runtime/storageNG/reference-mode-store.ts [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+      this.receiveQueue.push({...entry, promise: resolve});
+      if (startProcessing) {
+        void this.processQueue();
+      }
+    });
+  }
+
+  private async processQueue() {
+    while (this.receiveQueue.length > 0) {
+      const nextMessage = this.receiveQueue[0];
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: FYI you can do this instead: `const nextMessage = this.receiveQueue.shift();` THis way there's no danger of forgetting to shift the array at the end of the loop.

File: src/runtime/storageNG/reference-mode-store.ts [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+   * on a version as well as presence.
+   * 
+   * Sync requests are propagated upwards to the storage proxy.
+   */
+  private async handleContainerStore(message: ProxyMessage<ReferenceContainer>) {
+    switch (message.type) {
+      case ProxyMessageType.Operations: {
+        for (const operation of message.operations) {
+          const reference = this.operationElement<Reference>(operation);
+          let getEntity;
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Can you add a type for this? I think it might be: `let getEntity: () => (Entity | null);`

File: src/runtime/storageNG/reference-mode-store.ts [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+    return true;
+  }
+
+  /**
+   * Handle an update from an upstream StorageProxy.
+   * 
+   * Model and Operation updates apply first to the backing store, then to the container store.
+   * Backing store updates should never fail as updates are locally generated.
+   * For Operations:
+   * - If the container store update has succeeds, then the update is mirrored to non-sending StorageProxies.
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: "has" -> ""

File: src/runtime/storageNG/reference-mode-store.ts [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+   * 
+   * Sync requests are handled by directly constructing and sending a model 
+   */
+  private async handleProxyMessage(message: ProxyMessage<Container>): Promise<boolean> {
+    switch (message.type) {
+      case ProxyMessageType.Operations: {
+        const operations = message.operations;
+        for (const operation of operations) {
+          const entity = this.operationElement<Entity>(operation);
+          let reference = null;
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Type?

File: src/runtime/storageNG/reference-mode-store.ts [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+        if (pendingIds.length === 0) {
+          this.enqueueSend(send);
+        } else {
+          this.enqueueBlockingSend(pendingIds, send);
+        }
+        
+        break;
+      }
+      default:
+          throw new Error('Unexpected ProxyMessageType');
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Fix indent

File: src/runtime/storageNG/reference-mode-store.ts [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+      this.pendingSends.push({fn: runnable});
+    }
+  }
+
+  /**
+   * Enqueues a send function on the send queue, deferring execution until the
+   * provided id list is available in the backing store.
+   */
+  private enqueueBlockingSend(ids: string[], runnable: Runnable) {
+    const block = Math.random() + '';
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Would a counter work better here instead? There doesn't seem to be a need for this to be a random number, just unique.

File: src/runtime/storageNG/reference-mode-store.ts [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+        maxVersion = Math.max(maxVersion, entityVersion[key]);
+        model.singletons[key].values = {[entity[key].id]: {value: entity[key], version}};
+        model.singletons[key].version = version;
+      } else if (model.collections[key]) {
+        model.collections[key].values = {};
+        if (entityVersion[key] == undefined) {
+          entityVersion[key] = 0;
+        }
+        const version = {[this.crdtKey]: ++entityVersion[key]};
+        maxVersion = Math.max(maxVersion, entityVersion[key]);
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Lines 405-409 looks duplicated from the if-statement above. Could you hoist them outside the if-statements?

You might have an issue when `key === id` (you don't want to bump versions for that), but that is probably better dealt with explicitly anyway by adding `if (key === id) continue;` at the top of the for loop -- that seems clearer to me.

File: src/runtime/storageNG/reference-mode-store.ts [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+    }
+    for (const key of Object.keys(data.collections)) {
+      entity[key] = data.collections[key];
+    }
+    return entity;
+  }
+
+  private cloneMap<V>(map: Dictionary<V>) {
+    const result: Dictionary<V> = {};
+    Object.keys(map).forEach(key => result[key] = map[key]);
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Object.entries would let you avoid having to do the map lookup again

File: src/runtime/storageNG/reference-mode-store.ts [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+      }
+      void this.callbacks.get(key)({...message, id: key} as ProxyMessage<Container>);
+    }
+  }
+
+  /**
+   * Write the provided entity to the backing store.
+   */
+  private async updateBackingStore(entity: Entity) {
+    return this.backingStore.onProxyMessage({type: ProxyMessageType.ModelUpdate, model: this.entityToModel(entity).model, id: 1}, entity.id);
```
Comment (mykmartin) [inline src/runtime/storageNG/reference-mode-store.ts]: Nit: maybe multi-line this? It's a bit hard to parse on a quick read..

File: src/runtime/storageNG/reference-mode-store.ts [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+    }
+    this.queue[id] = [];
+  }
+}
+
+function isCollectionOperation<T>(operation: CollectionOperation<T> | SingletonOperation<T>): operation is CollectionOperation<T> {
+  if (operation['added'] || operation['removed']) {
+    return true;
+  }
+  return false;
```
Comment (mykmartin) [inline src/runtime/storageNG/reference-mode-store.ts]: `return Boolean(operation['added'] || operation['removed']);` ?

File: src/runtime/storageNG/reference-mode-store.ts [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+}
+
+function isSingletonOperation<T>(operation: CollectionOperation<T> | SingletonOperation<T>): operation is SingletonOperation<T> {
+  if (operation['added'] || operation['removed']) {
+    return false;
+  }
+  return true;
```
Comment (mykmartin) [inline src/runtime/storageNG/reference-mode-store.ts]: Ditto.

File: src/runtime/storageNG/tests/reference-mode-store-test.ts [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+  /*
+  it('will construct Direct stores when required', async () => {
+    DriverFactory.register(new MockStorageDriverProvider());
+
+    const store = new Store(testKey, Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
+    const activeStore = await store.activate();
+
+    assert.equal(activeStore.constructor, DirectStore);
+  });
+  */
```
Comment (mykmartin) [inline src/runtime/storageNG/tests/reference-mode-store-test.ts]: Did you mean to leave these commented out?

File: src/runtime/storageNG/direct-store.ts [hunk 17]
```diff
@@ -84,17 +84,26 @@ export class DirectStore<T extends CRDTTypeRecord> extends ActiveStore<T> {
...
+        }
+      });
     }
   }
 
-  private async processModelChange(modelChange: CRDTChange<T>, otherChange: CRDTChange<T>, version: number, fromDriver: boolean) {
-    this.deliverCallbacks(modelChange);
-    await this.updateStateAndAct(this.noDriverSideChanges(modelChange, otherChange, fromDriver), version, fromDriver);
+  private async processModelChange(modelChange: CRDTChange<T>, otherChange: CRDTChange<T>, version: number, channel: number) {
+    this.deliverCallbacks(modelChange, false, channel);
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/direct-store.ts]: done

File: src/runtime/storageNG/reference-mode-store.ts (hunk 5 above)
Comment (shans) [reply src/runtime/storageNG/reference-mode-store.ts]: Yeah, that's correct; can't shift until the end because otherwise the signalling gets messed up. I'll definitely add a comment here.

File: src/runtime/storageNG/reference-mode-store.ts (hunk 11 above)
Comment (shans) [reply src/runtime/storageNG/reference-mode-store.ts]: Yup! Moving maxVersion updating outside the if/then works.

File: src/runtime/storageNG/reference-mode-store.ts [hunk 18]
```diff
@@ -0,0 +1,598 @@
...
+    this.holdQueue.enqueue(ids, () => this.processPendingSends(block));
+  }
+
+  /**
+   * Process any sends in the pending send queue, including sends blocked on the
+   * provided block. This should only be called by the holdQueue.
+   */
+  private processPendingSends(block: string) {
+    while (this.pendingSends.length > 0) {
+      if (this.pendingSends[0].block == null || this.pendingSends[0].block === block) {
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/reference-mode-store.ts]: I'm following lint practice here - == null, === everything else

File: src/runtime/storageNG/tests/reference-mode-store-test.ts (hunk 16 above)
Comment (shans) [reply src/runtime/storageNG/tests/reference-mode-store-test.ts]: I should skip them instead. I haven't set up the construction pathway for this yet and these tests rely on that being there.

File: src/runtime/storageNG/tests/reference-mode-store-test.ts [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+  beforeEach(() => {
+    testKey = new MockHierarchicalStorageKey();
+    DriverFactory.clearRegistrationsForTesting();
+  });
+
+  after(() => {
+    DriverFactory.clearRegistrationsForTesting();
+  });
+
+  it.skip(`will throw an exception if an appropriate driver can't be found`, async () => {
```
Comment (csilvestrini) [inline src/runtime/storageNG/tests/reference-mode-store-test.ts]: How come this is skipped?

File: src/runtime/storageNG/tests/reference-mode-store-test.ts [hunk 20]
```diff
@@ -0,0 +1,403 @@
...
+    const store = new Store(testKey, Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
+    try {
+      await store.activate();
+      assert.fail('store.activate() should not have succeeded');
+    } catch (e) {
+      assert.match(e.toString(), /No driver exists/);
+    }
+  });
+
+  it.skip('will construct Direct stores when required', async () => {
```
Comment (csilvestrini) [inline src/runtime/storageNG/tests/reference-mode-store-test.ts]: Same here?

File: src/runtime/storageNG/reference-mode-store.ts [hunk 21]
```diff
@@ -0,0 +1,637 @@
...
+      singletons[field] = new CRDTSingleton();
+    }
+    const collections = {};
+    for (const field of Object.keys(model.collections)) {
+      collections[field] = new CRDTCollection();
+    }
+    const entityCRDT = new CRDTEntity(singletons, collections);
+    entityCRDT.merge(model);
+    const data = entityCRDT.getParticleView();
+    for (const key of Object.keys(data.singletons)) {
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: Optional: Object.entries would work well here too

File: src/runtime/storageNG/reference-mode-store.ts [hunk 22]
```diff
@@ -0,0 +1,637 @@
...
+      entity[key] = data.singletons[key];
+    }
+    for (const key of Object.keys(data.collections)) {
```
Comment (csilvestrini) [inline src/runtime/storageNG/reference-mode-store.ts]: And here

File: src/runtime/storageNG/tests/reference-mode-store-test.ts (hunk 19 above)
Comment (shans) [reply src/runtime/storageNG/tests/reference-mode-store-test.ts]: They don't currently make sense; as I figure out the construction story for referenceMode I'll unskip them.
